File: crates/globstar/src/engine/segment/compile.rs
use crate::engine::ops::Op;

use super::seg_nfa::SegNfa;
use super::{Elem, ElemSeq, MAX_FORKS, MAX_SEQ_STATES, Wild, WildKind};

pub(super) fn compile_seqs(ops: &[Op], dot: bool, ci: bool) -> Option<Vec<ElemSeq>> {
    if !ops
        .iter()
        .any(|op| matches!(op, Op::Alternation(_)) && op_crosses_segment(op))
    {
        return Some(vec![segmentize_fork(ops, dot, ci)?]);
    }
    let op_seqs = expand_forks(ops)?;
    let mut seqs = Vec::with_capacity(op_seqs.len());
    for fork in &op_seqs {
        seqs.push(segmentize_fork(fork, dot, ci)?);
    }
    Some(seqs)
}

fn segmentize_fork(ops: &[Op], dot: bool, ci: bool) -> Option<ElemSeq> {
    let glued = ops.windows(2).any(|w| {
        matches!(w[1], Op::GlobstarAny) && matches!(w[0], Op::OptSegmentsSlash | Op::SepRun)
    });
    if !glued {
        return segmentize(ops, dot, ci);
    }
    let mut flat = ops.to_vec();
    collapse_open_globstars(&mut flat);
    segmentize(&flat, dot, ci)
}

fn collapse_open_globstars(ops: &mut Vec<Op>) {
    let mut i = 0;
    while i < ops.len() {
        if i > 0 && matches!(ops[i], Op::GlobstarAny) {
            match ops[i - 1] {
                Op::OptSegmentsSlash => {
                    ops.remove(i - 1);
                    i -= 1;
                    continue;
                }
                Op::SepRun => {
                    ops[i - 1] = Op::SlashAnything;
                    ops.remove(i);
                    continue;
                }
                _ => {}
            }
        }
        i += 1;
    }
}

fn expand_forks(ops: &[Op]) -> Option<Vec<Vec<Op>>> {
    let mut seqs: Vec<Vec<Op>> = vec![Vec::with_capacity(ops.len())];
    for op in ops {
        if matches!(op, Op::Alternation(_)) && op_crosses_segment(op) {
            let Op::Alternation(branches) = op else {
                unreachable!()
            };
            let mut expanded: Vec<Vec<Op>> = Vec::new();
            for branch in branches {
                let sub = expand_forks(branch)?;
                expanded.extend(sub);
                if expanded.len() > MAX_FORKS {
                    return None;
                }
            }
            let mut next = Vec::with_capacity(seqs.len().saturating_mul(expanded.len()));
            for seq in &seqs {
                for exp in &expanded {
                    if next.len() >= MAX_FORKS {
                        return None;
                    }
                    let mut merged = seq.clone();
                    merged.extend(exp.iter().cloned());
                    next.push(merged);
                }
            }
            seqs = next;
        } else {
            for seq in seqs.iter_mut() {
                seq.push(op.clone());
            }
        }
    }
    Some(seqs)
}

fn op_crosses_segment(op: &Op) -> bool {
    match op {
        Op::Sep
        | Op::SepRun
        | Op::Globstar
        | Op::OptSegmentsSlash
        | Op::SlashAnything
        | Op::GlobstarAny
        | Op::LeadingSeps => true,
        Op::Alternation(branches) => branches.iter().any(|b| b.iter().any(op_crosses_segment)),
        _ => false,
    }
}

#[derive(PartialEq, Clone, Copy)]
enum Boundary {
    /// Sequence start, or right after a globstar element.
    Fresh,
    /// Just consumed a `Sep`.
    Strict,
    /// Just consumed a `SepRun` (lenient `**` boundary).
    Lenient,
    /// A `GlobstarAny`/`SlashAnything` absorbs the rest of the path;
    /// any op after it makes the sequence non-expressible.
    Open,
}

fn segmentize(ops: &[Op], dot: bool, ci: bool) -> Option<ElemSeq> {
    let mut elems: Vec<Elem> = Vec::with_capacity(8);
    let mut buf: Vec<Op> = Vec::new();
    let mut state = Boundary::Fresh;

    for (i, op) in ops.iter().enumerate() {
        if state == Boundary::Open {
            return None;
        }
        match op {
            Op::Lit(_) | Op::AnyChar | Op::Star | Op::Class(_) | Op::Alternation(_) => {
                if lit_contains_sep(op) {
                    return None;
                }
                push_in_seg(&mut buf, op);
            }
            Op::Sep => {
                elems.push(close_segment(&mut buf, dot, ci)?);
                state = Boundary::Strict;
            }
            Op::SepRun => {
                elems.push(close_segment(&mut buf, dot, ci)?);
                state = Boundary::Lenient;
            }
            Op::LeadingSeps => {
                if i != 0 {
                    return None;
                }
            }
            Op::OptSegmentsSlash => {
                if !buf.is_empty() {
                    return None;
                }
                let strict_entry = match state {
                    Boundary::Fresh => !elems.is_empty(),
                    Boundary::Strict => true,
                    Boundary::Lenient => false,
                    Boundary::Open => unreachable!("handled at the loop top"),
                };
                elems.push(if strict_entry {
                    Elem::G0Strict
                } else {
                    Elem::G0
                });
                state = Boundary::Fresh;
            }
            Op::SlashAnything => {
                elems.push(close_segment(&mut buf, dot, ci)?);
                elems.push(Elem::G1);
                state = Boundary::Open;
            }
            Op::GlobstarAny => {
                if !buf.is_empty() {
                    return None;
                }
                let strict = state == Boundary::Strict;
                elems.push(if strict { Elem::G1 } else { Elem::G0 });
                state = Boundary::Open;
            }
            Op::Globstar => return None,
        }
    }

    if state != Boundary::Open {
        elems.push(close_segment(&mut buf, dot, ci)?);
    }
    finish(elems)
}

fn lit_contains_sep(op: &Op) -> bool {
    match op {
        Op::Lit(bytes) => bytes.iter().any(|&b| std::path::is_separator(b as char)),
        Op::Alternation(branches) => branches.iter().any(|b| b.iter().any(lit_contains_sep)),
        _ => false,
    }
}

fn push_in_seg(buf: &mut Vec<Op>, op: &Op) {
    if let (Op::Lit(bytes), Some(Op::Lit(prev))) = (op, buf.last_mut()) {
        prev.extend_from_slice(bytes);
        return;
    }
    buf.push(op.clone());
}

fn close_segment(buf: &mut Vec<Op>, dot: bool, ci: bool) -> Option<Elem> {
    if buf.is_empty() {
        return Some(Elem::Lit(Box::from(&b""[..])));
    }
    let mut ops = std::mem::take(buf);
    if ops.len() == 1 {
        if let Op::Lit(bytes) = &mut ops[0] {
            return Some(Elem::Lit(std::mem::take(bytes).into_boxed_slice()));
        }
    }
    Some(Elem::Wild(compile_wild(&ops, dot, ci)?))
}

fn compile_wild(ops: &[Op], dot: bool, ci: bool) -> Option<Wild> {
    let mut idx = 0;
    let prefix: &[u8] = match ops.first() {
        Some(Op::Lit(b)) => {
            idx = 1;
            b
        }
        _ => b"",
    };
    let mut anychars = 0u32;
    let mut has_star = false;
    while idx < ops.len() {
        match &ops[idx] {
            Op::Star => has_star = true,
            Op::AnyChar => anychars += 1,
            _ => break,
        }
        idx += 1;
    }
    let has_wilds = has_star || anychars > 0;
    let dot_protect = !dot && prefix.is_empty() && has_wilds;

    if idx == ops.len() {
        return Some(Wild {
            kind: WildKind::Affix {
                prefix: Box::from(prefix),
                suffix: Box::from(&b""[..]),
            },
            min_len: prefix.len() as u32 + anychars,
            variable: has_star,
            dot_protect,
        });
    }
    if let Some(suffixes) = suffix_product(&ops[idx..]) {
        if suffixes.len() == 1 {
            let suffix = suffixes.into_iter().next().unwrap();
            return Some(Wild {
                min_len: (prefix.len() + suffix.len()) as u32 + anychars,
                kind: WildKind::Affix {
                    prefix: Box::from(prefix),
                    suffix: suffix.into_boxed_slice(),
                },
                variable: has_star,
                dot_protect,
            });
        }
        return Some(Wild {
            kind: WildKind::AffixSet {
                prefix: Box::from(prefix),
                suffixes: suffixes
                    .into_iter()
                    .map(Vec::into_boxed_slice)
                    .collect::<Vec<_>>()
                    .into_boxed_slice(),
            },
            min_len: prefix.len() as u32 + anychars,
            variable: has_star,
            dot_protect,
        });
    }
    let nfa = SegNfa::compile(ops, dot, ci)?;
    let dot_protect = !dot && nfa.wild_led;
    Some(Wild {
        kind: WildKind::Generic(nfa),
        min_len: 0,
        variable: true,
        dot_protect,
    })
}

const MAX_SUFFIX_PRODUCT: usize = 16;

fn suffix_product(ops: &[Op]) -> Option<Vec<Vec<u8>>> {
    if let [Op::Lit(bytes)] = ops {
        return Some(vec![bytes.clone()]);
    }
    let mut parts: Vec<Vec<u8>> = vec![Vec::new()];
    for op in ops {
        match op {
            Op::Lit(bytes) => {
                for p in parts.iter_mut() {
                    p.extend_from_slice(bytes);
                }
            }
            Op::Alternation(branches) => {
                let mut lits = Vec::with_capacity(branches.len());
                for b in branches {
                    match b.as_slice() {
                        [] => lits.push(&b""[..]),
                        [Op::Lit(bytes)] => lits.push(bytes.as_slice()),
                        _ => return None,
                    }
                }
                if parts.len() * lits.len() > MAX_SUFFIX_PRODUCT {
                    return None;
                }
                let mut next = Vec::with_capacity(parts.len() * lits.len());
                for p in &parts {
                    for l in &lits {
                        let mut v = Vec::with_capacity(p.len() + l.len());
                        v.extend_from_slice(p);
                        v.extend_from_slice(l);
                        next.push(v);
                    }
                }
                parts = next;
            }
            _ => return None,
        }
    }
    Some(parts)
}

fn finish(elems: Vec<Elem>) -> Option<ElemSeq> {
    let m = elems.len();
    let mut state_of = Vec::with_capacity(m);
    let mut n: usize = 0;
    for e in &elems {
        state_of.push(n as u8);
        n += match e {
            Elem::G0Strict | Elem::G1 => 2,
            _ => 1,
        };
        if n >= MAX_SEQ_STATES {
            return None;
        }
    }
    let accept = n;
    n += 1;

    let mut elem_of = vec![0u8; n];
    for (i, &entry) in state_of.iter().enumerate() {
        let end = if i + 1 < m {
            state_of[i + 1] as usize
        } else {
            accept
        };
        elem_of[entry as usize..end].fill(i as u8);
    }

    let mut eps: Vec<u64> = (0..n).map(|s| 1u64 << s).collect();
    for i in (0..m).rev() {
        let s = state_of[i] as usize;
        let next_entry = if i + 1 < m {
            state_of[i + 1] as usize
        } else {
            accept
        };
        match elems[i] {
            Elem::G0 => eps[s] |= eps[next_entry],
            Elem::G0Strict => {
                eps[s] |= eps[next_entry];
                eps[s + 1] |= eps[next_entry];
            }
            Elem::G1 => {
                eps[s + 1] |= eps[next_entry];
            }
            _ => {}
        }
    }

    let mut reach1: u64 = 0;
    let mut sat_tail = true;
    for i in (0..m).rev() {
        let sat_i = match &elems[i] {
            Elem::Wild(w) => match &w.kind {
                WildKind::Generic(nfa) => nfa.satisfiable,
                _ => true,
            },
            _ => true,
        };
        let s = state_of[i] as usize;
        let can = match &elems[i] {
            Elem::G0 | Elem::G0Strict | Elem::G1 => sat_tail,
            Elem::Lit(_) | Elem::Wild(_) => sat_i && sat_tail,
        };
        if can {
            reach1 |= 1u64 << s;
            if matches!(elems[i], Elem::G0Strict | Elem::G1) {
                reach1 |= 1u64 << (s + 1);
            }
        }
        sat_tail = sat_i && sat_tail;
    }

    let g_count = elems.iter().filter(|e| e.is_globstar()).count() as u8;
    let single_g = if g_count == 1 {
        elems.iter().position(Elem::is_globstar).unwrap() as u8
    } else {
        u8::MAX
    };

    let mut joined_head = Vec::new();
    if g_count == 1 && single_g > 0 {
        let head = &elems[..single_g as usize];
        if head.iter().all(|e| matches!(e, Elem::Lit(_))) {
            for e in head {
                if let Elem::Lit(bytes) = e {
                    joined_head.extend_from_slice(bytes);
                    joined_head.push(b'/');
                }
            }
        }
    }

    let quick_suffix: Box<[u8]> = match elems.last() {
        Some(Elem::Lit(bytes)) => bytes.clone(),
        Some(Elem::Wild(w)) => match &w.kind {
            WildKind::Affix { suffix, .. } => suffix.clone(),
            _ => Box::from(&b""[..]),
        },
        _ => Box::from(&b""[..]),
    };

    Some(ElemSeq {
        elems: elems.into_boxed_slice(),
        single_g,
        g_count,
        joined_head: joined_head.into_boxed_slice(),
        state_of: state_of.into_boxed_slice(),
        elem_of: elem_of.into_boxed_slice(),
        num_states: n as u8,
        eps: eps.into_boxed_slice(),
        reach1,
        quick_suffix,
    })
}

File: packages/globstar/src/engine/segment/compile.js
// Ops → element sequences: fork expansion, the segmentizer, and
// in-segment wildcard classification. JS port of the Rust module
// `crates/globstar/src/engine/segment/compile.rs`.

import {
  OP_LIT,
  OP_ANYCHAR,
  OP_STAR,
  OP_CLASS,
  OP_SEP,
  OP_SEP_RUN,
  OP_GLOBSTAR,
  OP_OPT_SEGMENTS_SLASH,
  OP_SLASH_ANYTHING,
  OP_GLOBSTAR_ANY,
  OP_LEADING_SEPS,
  OP_ALTERNATION,
  SLASH_ANY_OP,
} from "../ops/index.js";
import { CI_BYTE } from "../../ast.js";
import { isPathSep } from "../../options.js";
import { latin1 } from "../../utf8.js";
import {
  EL_LIT,
  EL_WILD,
  EL_G0,
  EL_G0_STRICT,
  EL_G1,
  WK_AFFIX,
  WK_AFFIX_SET,
  WK_GENERIC,
  MAX_FORKS,
  MAX_SEQ_STATES,
} from "./index.js";
import { SegNfa } from "./seg-nfa.js";

const MAX_SUFFIX_PRODUCT = 16;

function makeElem(kind, litBytes, wild) {
  return { kind, litStr: litBytes !== null ? latin1(litBytes) : null, wild };
}

function opCrossesSegment(op) {
  switch (op.kind) {
    case OP_SEP:
    case OP_SEP_RUN:
    case OP_GLOBSTAR:
    case OP_OPT_SEGMENTS_SLASH:
    case OP_SLASH_ANYTHING:
    case OP_GLOBSTAR_ANY:
    case OP_LEADING_SEPS:
      return true;
    case OP_ALTERNATION:
      for (const b of op.branches) {
        for (const o of b) if (opCrossesSegment(o)) return true;
      }
      return false;
    default:
      return false;
  }
}

function opIsCrossingAlt(op) {
  return op.kind === OP_ALTERNATION && opCrossesSegment(op);
}

function hasOpenGlobstarAdjacency(ops) {
  for (let i = 1; i < ops.length; i++) {
    if (
      ops[i].kind === OP_GLOBSTAR_ANY &&
      (ops[i - 1].kind === OP_OPT_SEGMENTS_SLASH || ops[i - 1].kind === OP_SEP_RUN)
    ) {
      return true;
    }
  }
  return false;
}

// `(?:[^/]*/)* .*` = `.*` → GlobstarAny; `/+ .*` = `/.*` → SlashAnything.
// Both language-preserving. Without them the segmentizer turns the `**`
// fork of `x/{**,a}/**` (`SepRun OSS GlobstarAny`) into `[Lit, G0, G0]`,
// dropping the mandatory separator and matching `x` (§8.3). Mutates and
// returns `ops` (caller passes a copy).
function collapseOpenGlobstars(ops) {
  let i = 0;
  while (i < ops.length) {
    if (i > 0 && ops[i].kind === OP_GLOBSTAR_ANY) {
      const prev = ops[i - 1].kind;
      if (prev === OP_OPT_SEGMENTS_SLASH) {
        ops.splice(i - 1, 1); // drop OSS, re-examine GlobstarAny
        i -= 1;
        continue;
      }
      if (prev === OP_SEP_RUN) {
        ops[i - 1] = SLASH_ANY_OP;
        ops.splice(i, 1);
        continue;
      }
    }
    i += 1;
  }
  return ops;
}

function expandForks(ops) {
  let crossing = false;
  for (const op of ops) {
    if (opIsCrossingAlt(op)) {
      crossing = true;
      break;
    }
  }
  if (!crossing) return [ops];
  let seqs = [[]];
  for (const op of ops) {
    if (opIsCrossingAlt(op)) {
      const expanded = [];
      for (const branch of op.branches) {
        const sub = expandForks(branch);
        if (sub === null) return null;
        for (const s of sub) expanded.push(s);
        if (expanded.length > MAX_FORKS) return null;
      }
      const next = [];
      for (const seq of seqs) {
        for (const exp of expanded) {
          if (next.length >= MAX_FORKS) return null;
          next.push(seq.concat(exp));
        }
      }
      seqs = next;
    } else {
      for (const seq of seqs) seq.push(op);
    }
  }
  return seqs;
}

export function opsHaveNonAscii(ops) {
  for (const op of ops) {
    switch (op.kind) {
      case OP_LIT:
        for (let i = 0; i < op.bytes.length; i++) {
          if (op.bytes[i] > 0x7f) return true;
        }
        break;
      case OP_CLASS:
        for (const it of op.cls.items) {
          if (it.tag === CI_BYTE ? it.b > 0x7f : it.hi > 0x7f) return true;
        }
        break;
      case OP_ALTERNATION:
        for (const b of op.branches) {
          if (opsHaveNonAscii(b)) return true;
        }
        break;
      default:
        break;
    }
  }
  return false;
}

function litContainsSep(op) {
  if (op.kind === OP_LIT) {
    for (let i = 0; i < op.bytes.length; i++) {
      if (isPathSep(op.bytes[i])) return true;
    }
    return false;
  }
  if (op.kind === OP_ALTERNATION) {
    for (const b of op.branches) {
      for (const o of b) if (litContainsSep(o)) return true;
    }
  }
  return false;
}

// Boundary states while segmentizing.
const B_FRESH = 0;
const B_STRICT = 1;
const B_LENIENT = 2;
// An absorber whose op form does not self-delimit (GlobstarAny,
// SlashAnything) was just pushed as the last element. Only the Sep
// that closes it may follow.
const B_OPEN = 3;

const EMPTY_BYTES = new Uint8Array(0);

// Compile the lowered ops into fork sequences. `null` means not
// segment-expressible, so the caller falls back to the PikeVm (ports
// `compile_seqs` + `segmentize_fork` in engine/segment/compile.rs).
export function compileSeqs(ops, dot, ci) {
  const opSeqs = expandForks(ops);
  if (opSeqs === null) return null;
  const seqs = [];
  for (let fork of opSeqs) {
    // Collapse open-globstar adjacencies fork-splicing / separator
    // distribution can create, before segmentizing. Copy first — the
    // no-crossing path returns the caller's ops verbatim.
    if (hasOpenGlobstarAdjacency(fork)) fork = collapseOpenGlobstars(fork.slice());
    const seq = segmentize(fork, dot, ci);
    if (seq === null) return null;
    seqs.push(seq);
  }
  return seqs;
}

function segmentize(ops, dot, ci) {
  const elems = [];
  let buf = [];
  let state = B_FRESH;
  let leadingSeps = false;

  const closeSegment = () => {
    if (buf.length === 0) return makeElem(EL_LIT, EMPTY_BYTES, null);
    const segOps = buf;
    buf = [];
    if (segOps.length === 1 && segOps[0].kind === OP_LIT) {
      return makeElem(EL_LIT, segOps[0].bytes, null);
    }
    const wild = compileWild(segOps, dot, ci);
    return wild === null ? null : makeElem(EL_WILD, null, wild);
  };

  for (let i = 0; i < ops.length; i++) {
    const op = ops[i];
    // The only op that may follow an open absorber is the separator
    // closing its right edge, and it upgrades the lenient `.*` to "at
    // least one segment". A G1 absorber (SlashAnything, or GlobstarAny
    // behind a strict Sep) is never followed by a Sep after lowering;
    // bail rather than drop the separator.
    if (state === B_OPEN) {
      if (op.kind !== OP_SEP || elems[elems.length - 1]?.kind !== EL_G0) return null;
      elems[elems.length - 1] = makeElem(EL_G1, null, null);
      state = B_FRESH;
      continue;
    }
    switch (op.kind) {
      case OP_LIT:
      case OP_ANYCHAR:
      case OP_STAR:
      case OP_CLASS:
      case OP_ALTERNATION: {
        if (litContainsSep(op)) return null; // escaped separator
        pushInSeg(buf, op);
        break;
      }
      case OP_SEP: {
        const e = closeSegment();
        if (e === null) return null;
        elems.push(e);
        state = B_STRICT;
        break;
      }
      case OP_SEP_RUN: {
        // Generated only immediately before an OSS.
        const e = closeSegment();
        if (e === null) return null;
        elems.push(e);
        state = B_LENIENT;
        break;
      }
      case OP_LEADING_SEPS: {
        if (i !== 0) return null;
        leadingSeps = true;
        break;
      }
      case OP_OPT_SEGMENTS_SLASH: {
        // A glued absorber cannot be produced today (the parser degrades
        // any `**` that does not own a whole segment, §8.1). Defensive
        // bail, PikeVm answers correctly if one ever appears.
        if (buf.length > 0) return null;
        let strictEntry;
        if (state === B_FRESH) strictEntry = !leadingSeps && elems.length > 0;
        else if (state === B_STRICT) strictEntry = true;
        else strictEntry = false; // B_LENIENT
        elems.push(makeElem(strictEntry ? EL_G0_STRICT : EL_G0, null, null));
        state = B_FRESH;
        leadingSeps = false;
        break;
      }
      case OP_SLASH_ANYTHING: {
        // Trailing `/**`: brings its own leading boundary.
        const e = closeSegment();
        if (e === null) return null;
        elems.push(e);
        elems.push(makeElem(EL_G1, null, null));
        state = B_OPEN;
        break;
      }
      case OP_GLOBSTAR_ANY: {
        // Defensive bail, same as the OSS arm above.
        if (buf.length > 0) return null;
        // Behind a strict separator the absorber must consume >= 1
        // segment (`a/{**,x}` rejects `a`).
        const strict = state === B_STRICT;
        elems.push(makeElem(strict ? EL_G1 : EL_G0, null, null));
        state = B_OPEN;
        break;
      }
      default:
        return null; // raw OP_GLOBSTAR never survives the fold
    }
  }

  if (state !== B_OPEN) {
    const e = closeSegment();
    if (e === null) return null;
    elems.push(e);
  }
  return finishSeq(elems);
}

function pushInSeg(buf, op) {
  if (op.kind === OP_LIT && buf.length > 0 && buf[buf.length - 1].kind === OP_LIT) {
    const prev = buf[buf.length - 1];
    const merged = new Uint8Array(prev.bytes.length + op.bytes.length);
    merged.set(prev.bytes, 0);
    merged.set(op.bytes, prev.bytes.length);
    buf[buf.length - 1] = { kind: OP_LIT, bytes: merged };
    return;
  }
  buf.push(op);
}

// ---------------------------------------------------------------------------
// Wild classification
// ---------------------------------------------------------------------------

function makeWild(kind, fields) {
  return {
    kind,
    prefixStr: fields.prefixStr ?? "",
    suffixStr: fields.suffixStr ?? "",
    suffixSetStr: fields.suffixSetStr ?? null,
    minLen: fields.minLen ?? 0,
    variable: fields.variable ?? true,
    dotProtect: fields.dotProtect ?? false,
    anychars: fields.anychars ?? 0,
    nfa: fields.nfa ?? null,
  };
}

function compileWild(ops, dot, ci) {
  let idx = 0;
  let prefix = EMPTY_BYTES;
  if (ops[0].kind === OP_LIT) {
    prefix = ops[0].bytes;
    idx = 1;
  }
  let anychars = 0;
  let hasStar = false;
  while (idx < ops.length) {
    const k = ops[idx].kind;
    if (k === OP_STAR) hasStar = true;
    else if (k === OP_ANYCHAR) anychars++;
    else break;
    idx++;
  }
  const hasWilds = hasStar || anychars > 0;
  const dotProtect = !dot && prefix.length === 0 && hasWilds;

  if (idx === ops.length) {
    return makeWild(WK_AFFIX, {
      prefixStr: latin1(prefix),
      minLen: prefix.length + anychars,
      variable: hasStar,
      dotProtect,
      anychars,
    });
  }
  const suffixes = suffixProduct(ops, idx);
  if (suffixes !== null) {
    if (suffixes.length === 1) {
      return makeWild(WK_AFFIX, {
        prefixStr: latin1(prefix),
        suffixStr: latin1(suffixes[0]),
        minLen: prefix.length + suffixes[0].length + anychars,
        variable: hasStar,
        dotProtect,
        anychars,
      });
    }
    return makeWild(WK_AFFIX_SET, {
      prefixStr: latin1(prefix),
      suffixSetStr: suffixes.map(latin1),
      minLen: prefix.length + anychars,
      variable: hasStar,
      dotProtect,
      anychars,
    });
  }
  const nfa = SegNfa.compile(ops, dot, ci);
  if (nfa === null) return null;
  return makeWild(WK_GENERIC, {
    dotProtect: !dot && nfa.wildLed,
    nfa,
  });
}

function suffixProduct(ops, from) {
  // Overwhelmingly common tail: one literal op (`*.ts`).
  if (from + 1 === ops.length && ops[from].kind === OP_LIT) {
    return [ops[from].bytes];
  }
  let parts = [[]];
  for (let i = from; i < ops.length; i++) {
    const op = ops[i];
    if (op.kind === OP_LIT) {
      for (const p of parts) {
        for (let j = 0; j < op.bytes.length; j++) p.push(op.bytes[j]);
      }
    } else if (op.kind === OP_ALTERNATION) {
      const lits = [];
      for (const b of op.branches) {
        if (b.length === 0) lits.push(EMPTY_BYTES);
        else if (b.length === 1 && b[0].kind === OP_LIT) lits.push(b[0].bytes);
        else return null;
      }
      if (parts.length * lits.length > MAX_SUFFIX_PRODUCT) return null;
      const next = [];
      for (const p of parts) {
        for (const l of lits) {
          const v = p.slice();
          for (let j = 0; j < l.length; j++) v.push(l[j]);
          next.push(v);
        }
      }
      parts = next;
    } else {
      return null;
    }
  }
  return parts.map((p) => Uint8Array.from(p));
}

// ---------------------------------------------------------------------------
// Element-NFA metadata (ports `finish` in engine/compile.rs)
// ---------------------------------------------------------------------------

function finishSeq(elems) {
  const m = elems.length;
  const stateOf = [];
  let n = 0;
  for (const e of elems) {
    stateOf.push(n);
    n += e.kind === EL_G0_STRICT || e.kind === EL_G1 ? 2 : 1;
    if (n >= MAX_SEQ_STATES) return null;
  }
  const accept = n;
  n += 1;

  // Inverse map: owning element per state (accept slot unused).
  const elemOf = new Array(n).fill(0);
  for (let i = 0; i < m; i++) {
    const end = i + 1 < m ? stateOf[i + 1] : accept;
    for (let s = stateOf[i]; s < end; s++) elemOf[s] = i;
  }

  const eps = new Array(n);
  for (let s = 0; s < n; s++) eps[s] = 1 << s;
  for (let i = m - 1; i >= 0; i--) {
    const s = stateOf[i];
    const nextEntry = i + 1 < m ? stateOf[i + 1] : accept;
    const k = elems[i].kind;
    if (k === EL_G0) eps[s] |= eps[nextEntry];
    else if (k === EL_G0_STRICT) {
      eps[s] |= eps[nextEntry];
      eps[s + 1] |= eps[nextEntry];
    } else if (k === EL_G1) {
      eps[s + 1] |= eps[nextEntry];
    }
  }

  const satFrom = new Array(m + 1).fill(true);
  for (let i = m - 1; i >= 0; i--) {
    satFrom[i] = elemSatisfiable(elems[i]) && satFrom[i + 1];
  }
  let reach1 = 0;
  for (let i = 0; i < m; i++) {
    const s = stateOf[i];
    const k = elems[i].kind;
    const isG = k === EL_G0 || k === EL_G0_STRICT || k === EL_G1;
    const can = isG ? satFrom[i + 1] : satFrom[i];
    if (can) {
      reach1 |= 1 << s;
      if (k === EL_G0_STRICT || k === EL_G1) reach1 |= 1 << (s + 1);
    }
  }

  let gCount = 0;
  let singleG = -1;
  for (let i = 0; i < m; i++) {
    const k = elems[i].kind;
    if (k === EL_G0 || k === EL_G0_STRICT || k === EL_G1) {
      gCount++;
      if (singleG === -1) singleG = i;
    }
  }
  if (gCount !== 1) singleG = -1;

  // Pre-join all-literal heads for the single-globstar fast path (mirrors
  // Rust `finish`). "src/**/.." heads become one prefix compare instead of a
  // segment loop. String form only. Byte mode is rare and keeps the loop.
  let joinedHeadStr = null;
  if (gCount === 1 && singleG > 0) {
    let allLit = true;
    for (let i = 0; i < singleG; i++) {
      if (elems[i].kind !== EL_LIT) {
        allLit = false;
        break;
      }
    }
    if (allLit) {
      let h = "";
      for (let i = 0; i < singleG; i++) h += elems[i].litStr + "/";
      joinedHeadStr = h;
    }
  }

  // Per-fork quick-reject suffix from the final element (only
  // consulted by multi-fork matchers).
  let quickStr = "";
  const lastEl = elems[m - 1];
  if (lastEl.kind === EL_LIT) quickStr = lastEl.litStr;
  else if (lastEl.kind === EL_WILD && lastEl.wild.kind === WK_AFFIX) {
    quickStr = lastEl.wild.suffixStr;
  }

  return {
    elems,
    singleG,
    gCount,
    stateOf,
    elemOf,
    numStates: n,
    eps,
    reach1,
    quickSuffixStr: quickStr,
    joinedHeadStr,
  };
}

function elemSatisfiable(e) {
  if (e.kind !== EL_WILD) return true;
  const w = e.wild;
  return w.kind === WK_GENERIC ? w.nfa.satisfiable : true;
}
